source/common/router/scoped_rds.cc

```diff
@@ -0,0 +1,272 @@
+      config_(std::make_shared<ThreadLocalScopedConfigImpl>(std::move(scope_key_builder))),
+      config_protos_(std::make_move_iterator(config_protos.begin()),
+                     std::make_move_iterator(config_protos.end())),
+      rds_config_source_(std::move(rds_config_source)) {}
+
+ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
+    const envoy::config::filter::network::http_connection_manager::v2::ScopedRds& scoped_rds,
+    const uint64_t manager_identifier, const std::string& name,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+    ScopedRoutesConfigProviderManager& config_provider_manager)
+    : DeltaConfigSubscriptionInstance(
+          "SRDS", manager_identifier, config_provider_manager, factory_context.timeSource(),
+          factory_context.timeSource().systemTime(), factory_context.localInfo()),
+      name_(name),
+      scope_(factory_context.scope().createScope(stat_prefix + "scoped_rds." + name + ".")),
+      stats_({ALL_SCOPED_RDS_STATS(POOL_COUNTER(*scope_))}) {
+  subscription_ = Envoy::Config::SubscriptionFactory::subscriptionFromConfigSource(
+      scoped_rds.scoped_rds_config_source(), factory_context.localInfo(),
+      factory_context.dispatcher(), factory_context.clusterManager(), factory_context.random(),
+      *scope_, "envoy.api.v2.ScopedRoutesDiscoveryService.FetchScopedRoutes",
+      "envoy.api.v2.ScopedRoutesDiscoveryService.StreamScopedRoutes",
+      Grpc::Common::typeUrl(
+          envoy::api::v2::ScopedRouteConfiguration().GetDescriptor()->full_name()),
+      factory_context.api());
+}
+
+void ScopedRdsConfigSubscription::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const std::string& version_info) {
+  if (resources.empty()) {
+    ENVOY_LOG(debug, "Empty resources in scoped RDS onConfigUpdate()");
+    stats_.update_empty_.inc();
+    ConfigSubscriptionCommonBase::onConfigUpdateFailed();
+    return;
+  }
+  std::vector<envoy::api::v2::ScopedRouteConfiguration> scoped_routes;
+  for (const auto& resource_any : resources) {
+    scoped_routes.emplace_back(
+        MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(resource_any));
+  }
+
+  std::unordered_set<std::string> resource_names;
+  for (const auto& scoped_route : scoped_routes) {
+    if (!resource_names.insert(scoped_route.name()).second) {
+      throw EnvoyException(
+          fmt::format("duplicate scoped route configuration {} found", scoped_route.name()));
+    }
+  }
+  for (const auto& scoped_route : scoped_routes) {
+    MessageUtil::validate(scoped_route);
+  }
+
+  // TODO(AndresGuedez): refactor such that it can be shared with other delta APIs (e.g., CDS).
+  std::vector<std::string> exception_msgs;
+  // We need to keep track of which scoped routes we might need to remove.
+  ScopedConfigManager::ScopedRouteMap scoped_routes_to_remove =
+      scoped_config_manager_.scopedRouteMap();
```

Comment:
Is the state kept in scoped_config_manager_ used anywhere else (or are there plans for that)? If not, I think you could clear the scoped route map on each update and then add routes that came in onConfigUpdate...

Reply:
The state must persist across `onConfigUpdate()` calls such that the diff can be computed between the previous set of `ScopedRouteConfiguration` protos and the new set of protos.